### ISSUE TITLE
RPI3 won't boot now - Problem with 32 built OP-TEE builds???

### ISSUE THREAD

**Issue by dnovick** [OP-TEE/optee_os#1526]
So, I've been trying to get a crypto on TZ demo going on the RPI3. After working through a number of issues with the code I'm using, I finally got everything working. In order to do this, I had to move my secure world to 32 bit (problem with the math lib I'm using). All was well.

However, then I went and made some changes to my host app and ta...dumped them on the rootfs and tried to boot, only to see what appeared to be old images running. Very strange.

So, to rule out stupidity on my part, I cleared my SD card of all partitions...and went ahead and wrote 0s to the entire 32GB...and then started all over with a clean build and new image creation.

Now, the Pi3 won't boot at all..

After the "starting kernel..." line, I get 4 messages and then it just sits there. I have attached an image of my UART output. I tried another PI3, thinking I may have broken my device, but it behaves the same.

This is very strange...I have  done nothing to cause this to happen...and now it is happening right after I've finally gotten my demo working. I'm desperate for help.

Thanks,
Dave
![uart](https://cloud.githubusercontent.com/assets/4956601/26014834/30d9c142-3713-11e7-87d0-75d50ad5e7c0.jpg)

**Comment by dnovick** [comment]
Here is more information. So, I feared that possibly my problem was that my Linux VM was somehow corrupted and had started causing this problem. So, I built a new VM last night, installed all the tools, and rebuilt the image for the RPI3...first with 64 bit binaries. This worked fine. I then rebuilt with 32 bit binaries...and this failed with exactly the same behavior.

What has happened to 32 bit builds of the secure world? This worked for me at first...and now it doesn't work at all???

Dave

**Comment by igoropaniuk** [comment]
Hi @dnovick,
Actually, nowadays we don't support 32-bit builds of both OP-TEE OS and Linux kernel for RPI3 .
And, frankly speaking, no one even tried to build and run it.

> What has happened to 32 bit builds of the secure world? This worked for me at first...and now it doesn't work at all???

I may be wrong but I guess that you have used 64-bit binaries for the first time, that's why it worked (for example, for some reason you might forget to update optee.bin; or `make` process just failed and you copied old 64-bit binaries)

I'm not sure why do you need to have 32-bit of OP-TEE OS core, isn't it sufficient just to have `COMPILE_S_USER    ?= 32`? You can check Hikey configuration, where everything is 64-bit, except of user TAs 

Hope it helps

Best regards,
Igor

**Comment by dnovick** [comment]
I'm sorry if I haven't been clearer, but the only thing I am running 32 bit is the secure world. I have no need to run Linux in 32 bit mode, and already knew this wasn't supported. 

I also know that I didn't accidentally deploy old 64 bit binaries, as due to a math lib bug, my TA won't work under 64 bit. However, when I deployed 32 bit secure world, it worked initially. That's why I need to use 32 bit secure world.

**Comment by jforissier** [comment]
@dnovick are you 100% sure you didn't deploy a 32-bit TA over a 64-bit TEE kernel (which is perfectly acceptable) when it worked initially? Do you really need to run a 32-bit TEE *kernel*? (that's @igoropaniuk's question and you didn't really answer that part). Sorry for insisting but I'd like to avoid misunderstandings... I repeat: you *can* run a 32-bit TA on top of a 64-bit OP-TEE OS!

**Comment by dnovick** [comment]
So, first, I didn't realize that you can actually run a 32 bit TA over a 64 bit kernel. My apologies if I missed that.

As far as what my state was when I had success, I changed the 2 makefile params for the secure world in the top level Makefile, and built all...at least I thought I built all. I suppose it is possible that I only rebuilt my host and TA...and therefore was running a 64 bit kernel with 32 bit TA. If that's what happened, and that is supported, I'm going to jump on that. I was pretty sure I built all at first...but in the rush may have not...and therefore feel stupid. Given what you are saying about 32 bit TEE kernel support, it sounds like that may be what happened.

That leads me to a question. Does this mean that I should just override the secure world x-bit settings in my TA makefile to force 32 bit build of the TA? It seems hinky to me to build 64 bit everything, and then change the params in the top level Makefile and just build my TA code.

Thanks,
Dave

**Comment by dnovick** [comment]
Just to clarify, the makefile I'm referring to as "top level" is build/rpi3.mk

**Comment by jforissier** [comment]
Yes just change this line: https://github.com/OP-TEE/build/blob/master/rpi3.mk#L8, replacing 64 by 32 and build everything. The TAs will be 32 bits and everything else will be 64.

**Comment by dnovick** [comment]
Really? That's exactly what I did that caused the problem. Won't that build a 32 bit TEE kernel? I thought I should just leave that as 64 bit...and change my TA Makefile by adding the same lines and changing the secure world ones to 32.

**Comment by jforissier** [comment]
Really, yes.

**Comment by dnovick** [comment]
So, I just noticed that these lines show up in both build/Makefile AND build/rpi3.mk. I hadn't seen that previously. So, are you saying that I should leave all of those in build/Makefile as 64....and the two S_xxx ones in rpi3.mk as 32?

My apologies for any confusion on my part. I think what may have happened actually is that I ended up overriding the ones in Makefile....and this hung me. Can you confirm?

**Comment by dnovick** [comment]
I also notice in both places that there is a separate line for secure kernel vs secure user mode. As I understand it, the TA's run in secure user mode. Would it not then be acceptable to just set the COMPILE_S_USER to 32, and leave the other secure entry at 64?

Mind you, I'll do whatever brings me back to life :-).

**Comment by jforissier** [comment]
build/Makefile is supposed to be a symlink to build/rpi3.mk
Yes, TAs do run in user mode.

**Comment by vchong** [comment]
Steps to rebuild. Starting with a clean environment and a fresh pull of the code based on rpi3.xml or rpi3_stable.xml:

    Change the build/Makefile entries override COMPILE_S_USER to 32.
    Within the build dir, perform a make all.
    Build the SD card image according to the instructions found by executing make img-help.
    Move the SD card to the RPI3.
    Power up.

build/Makefile is just a symlink to build/rpi3.mk. If you edit one you'll the change in the other.

**Comment by dnovick** [comment]
Ok gotcha....I had moved my original build directory over to the new VM and was working on that to try to reconstruct history. It looks like the symlink was broken when I moved it from my other VM. Now that I think about it, I do recall Makefile being a symlink.

All right....so I just need to override the COMPILE_S_USER to 32....That makes sense. I suspect what happened at first is that I had built 64 bit....overrode both S_ settings, but only built my code...which had the same affect as just overriding the COMPILE_S_USER to 32 (in both files until I fix the symlink).

I'm building all now. I'll let you know what happens. 

Thanks so much for taking the time to help, especially since this looks self-inflicted. The fact that one should run 64 bit OP-TEE kernel along 32 bit TAs escaped me. From previous posts, I was under the impression that although 32 bit Linux wasn't supported, 32 bit OP-TEE kernel and TAs were...and both the secure kernel and TA needed to be the same. I sit corrected.

Dave

**Comment by dnovick** [comment]
OK...so my build is running fine now. However, my TA crashes now, just like it does under 64 bit TA build. 

This seems strange, as I know the math bug doesn't manifest itself in general 32 bit ARM testing. I also know that I did have success last Friday. I loaded the TA under IdaPro, and it is indeed a 32 bit binary.

I'm going to track down exactly where the failure is happening now. I'm guessing its the same place as in the 64 bit case, but need to confirm. I had done some work to see if I could workaround the issue in 64 bit mode, but hadn't succeeded. Something during code generation of a couple of specific single line statements that include multiple casts is going wrong. My initial efforts to workaround it by breaking the line up into multiple ones causes the crash to disappear, but creates incorrect results. If it turns out that it's a similar issue in 32 bit mode, it may be that I can create a proper workaround that cures the crash and still produces correct results. The other possibility is to compare the compile flags between the standalone test build of the crypto/math lib and those of my TA build. If there is something significantly different there, changing my build flags could cure both the 32 bit and 64 bit issues.

That said, I'm still at a complete loss as to what mixture of hand waving and genuflection caused things to work once. 

Anyway, I really appreciate the help in making some progress today.

Dave
